### DISCUSSION TITLE
Animate USD Prims within Maya

### DISCUSSION THREAD

**Discussion by christophezito** [Autodesk/maya-usd#2116] [Q&A]
Hello,

I've seen this question has already been asked in this [discussion topic](https://github.com/Autodesk/maya-usd/discussions/1320).

For setdress purposes or to quickly animate a prop in the stage (Open a door for example), it would be very interresting to be able to animate USD Prims within Maya.

We tried many things to achieve this goal and I wondered if Autodesk or any User have a better approach. 

## Custom Node
Implementing our custom node that will be placed between two `MayaUsdProxyShape` nodes, and allow us to interact with the selected `Prim` to animate it using Maya anim curves, and write time samples on the fly in the USD Stage.

We still have some issues with the `LayerManager`, as if a `MayaUsdProxyShape` node has an incoming Stage the incoming node will be in charge of saving the USD edits.

This might not be the better way to do that, as we will have to maintain our custom node as well as a patch of  Maya USD to let the leaf `MayaUsdProxyShape` node be  in charge of saving the edits if our custom node is connected as an incoming stage.

## Use ProxyAccessor to animate Prims

Using the `ProxyAcessor`, we can have access to `xfomOps` and animate the `Prims` straight into maya setting keys on both side at the same time (Maya keys to the ProxyAccessor's Access Plugs and TimeSamples to the USD Prim)

This Proove Of Concept might work, but we don't have access to Maya Anim curves, and the animation interpolation is linear (which is normal).

## Use ProxyAccessor Access Plugs connected to Maya Rig

We tried another alternative: connect all the transform outputs of a Locator to the corresponding ProxyAccessor Access Plug, and an Offset Parent Matrix connection between the Locator and the current Prim's parent (to preserve previous transformation in the graph).
We can bake the Animation, setting Maya keys to the ProxyAccessor Access Plug and adding the corresponding (per frame) Time Samples To the Prim.
This seems to work, and we still need to improve the workflow to retrieve an already animated Prim.

## Autodesk point of view

Does Autodesk have plans to make USD Prims animation within Maya?
I guess it would be great to be able to manipulate, animate USD Prims like any other DAG Object, having access to the keyframes on the timeline and in the graph editor, and letting Maya do all the heavy work to write keys on both sides (Maya and USD)

## Users point of view

Does anyone else tried to achieve this goal and has a better approach?

Thanks,

Christophe

**Comment by kxl-adsk** [comment]
The proxy accessor is the foundational piece we were planning to build on top of. I would be curious to learn more about the issues you run into with `Use ProxyAccessor to animate Prims`. What do you mean by saying that you don't have access to the curves?

**Comment by christophezito** [comment]
It's good to know that using the `ProxyAccessor` seems to be the way to animate USD Prims within Maya.

Let me detail the issue we get animating the USD Prims within Maya.
The workflow is as follow:
- Load the stage the usual way using a `MayaUsdProxyShape`
- Select a `Prim`, and apply a transformation
- Key the appropriate `xformOp`on both sides (Maya keyframes and USD Time Samples)
```python
from mayaUsd import lib as mayaUsdLib
import ufe


def keyUfeObject():
    ufeObject = pa.getUfeSelection()
    selDag, selPrim = pa.getDagAndPrimFromUfe(ufeObject)
    stage = mayaUsdLib.GetPrim(selDag).GetStage()
    
    primPath = Sdf.Path(selPrim)
    prim = stage.GetPrimAtPath(primPath)
    
    pa.createXformOps(ufeObject)
    i = cmds.currentTime(query=True)
    for usdAttr in ["xformOp:rotateXYZ", "xformOp:translate"]:
        xformop = prim.GetAttribute(usdAttr)
        pos = xformop.Get()
        """Add a time sample to the xformOp at frame i"""
        xformop.Set(time=i, value=tuple(pos))
        """Add a keyframe to the proxyAccessor plug at frame i"""
        pa.keyframeAccessPlug(ufeObject, usdAttr)
```

This seems to work as expected, and we can use the "Save USD Edits" callback that will let us choose to save the animation in the Maya Scene or in the USD file.

But there still are some limitations:
- The keyframes are not displayed on the timeline.
- To access the animation curves in the Graph Editor, you need to select the `MayaUsdProxyShape`, and all the animation curves are displayed (It can be a lot), you can't get only the animation curves of the selected Prim.
- The usual Maya channel box's contextual menu "Set Keyframe" has no effect.
- In order to match the keyframed animation (with tangents) and the time samples, we need to bake the time samples (linear interpolation) at each frame.

These are more limitations from a User point of view. And this actually is the only reason we tried to add a minimal rig on top of the prim to ease the user's experience.

I think the perfect behaviour when animating USD Prims would be:
- To animate USD Prims the same way you animate any DAG Object
- The channel box xformOp available would be keyable
- The Graph Editor would display the selected Prim's animation curves
- The timeline would display the selected Prim's keyframes
- Maya would add the corresponding Prim's time samples (per frame) when saving the USD file
- When reading an USD file with animated Prims (time samples), it would add the corresponding Plug Access keyframes (or interpret them)

But maybe is that what you already have in mind, or even a better approach. Or maybe is that already working out of the box and I  miss something in the proper way to animate USD Prims.

We would like to stick as close as possible to the direction you are going to build this feature on, and the simpler might be the better.

**Reply by kxl-adsk** [reply]
Thank you. The limitations you listed are matching my expectations. We provided a proxy accessor to enable studios to build on top of it, but we don't have yet many workflows that leverage it. The work that currently is landing in dev will take advantage of "parenting" dag nodes under prims...but everything that you listed in regards to animation curves driving prims and the workflows with it is still missing.

I think it's also worth bringing up that some of the workflows you are describing would really benefit from having an official animation curve schema in USD. Yet another big topic for the community to align on.

**Reply by bogdanamidzic** [reply]
Are there any news on this? It is very very sad how broken whole USD is in Maya. 2 years passed, and it seems nothing changed. We had project which we had to animate inside Houdini, because animating groups in Maya was rocket science. And now 2 years later, opening it and realizing nothing has changed. 

In Houdini you can bind USD prims directly to kinefx skeleton, transform pieces, whatever you want. You get these transforms simply written in USD.

**Comment by p-hong** [comment]
Hello,

We've been looking into migrating our USD pipe to maya2022. Our current maya2020 pipe involves using temporary rig setups created in-shot for USD assemblies. This is to simplify assembly selection, manipulation, and keying for animators. It also lightens up rigging's workload a lot. We also rely on animators being able to set keys on assembly hierarchies as they need, with or without a temporary rig. Transforms can then be baked out at publish.

So it was a big surprise to find out we cannot freely constraint or key USD asset hierarchies in maya2022. Setting keys and constraints through ```ProxyAccessor``` might be a good workaround, but is not feasible for shot departments due to the limitations and complexities mentioned above.

Thanks to the insights from @christophezito, we will explore connecting our temporary rigs through access plugs instead of constraints.

Thanks,
Paul

**Comment by richardHaw** [comment]
any updates?

**Comment by wallworm** [comment]
Right now there is a proposal from Pixar for animation curves. Our intent is to work on supporting this when official.

**Reply by bogdanamidzic** [reply]
Are you bothered by the fact that people are waiting for these features for 3+ years?

**Reply by wallworm** [reply]
Of course we would love to tackle everyone's need as quickly as possible. We would like to accelerate some work here for sure. There are some things we will consider in the coming months around Constraints that can help improve animation workflows. The animation curves directly on USD will add another layer of improvements on animating USD. The truth is that we have several gaps from asset creation to layout to animation to rendering, and we are trying to make significant inroads on all fronts. I can't speak for the past, but I can speak for what we are trying to do now.

**Reply by bogdanamidzic** [reply]
Any news on this?

**Reply by wallworm** [reply]
Nothing explicitly new but we are investigating opportunities to improve the way ProxyAccessor works.

**Reply by bogdanamidzic** [reply]
Here's how that works in Houdini:
https://youtu.be/blpCOCTKBHo
https://youtu.be/aLyO883mkoQ

**Reply by wallworm** [reply]
Thanks for sharing. We'd like to improve direct animation workflows for USD in Maya for sure. The ProxyAccessor paths mentioned above will likely play some role, but so will USD Anim. Since USD Anim is currently rolling out in USD, this will start to help us build workflows for more direct animation on USD. For constraining, we will want to improve the ProxyAccessor which may need some redesign to accommodate the kinds of rigs and scenes that customers have been sending us. This work is still on our radar.

**Reply by bogdanamidzic** [reply]
Guys, I'm again on the task of trying to incorporate Maya into our USD pipeline.
I have proposal, maybe Autodesk could hire me as consultant part time, I show you how that works in Houdini and how it should be working in Maya.

**Reply by wallworm** [reply]
We would be happy to meet with you and hear your thoughts. Feel free to reach out to me at shawn.olson@autodesk.com and we can talk about any topics related to USD and production.

**Comment by phyrn** [comment]
Hey all, 
I work in the film industry 
win maya 2024 mayausd 30

the workaround I have for animating an asset prim (referenced  asset usda) is creating a new stage with the usda from the asset directly. 
this way  I can key the asset with mayas default animation systems. when im done I convert the keys to timesamples in a sublayer.

this works really well , but i am having an issue with the stage not being at root level. 

if I have the stage grouped  under a maya transform the usd layer editor UI doesnt seem to work anymore. 

I think this is because the UI uses short paths when building its commands  but some of the functions require a longpath. 

do you know if there is any way around this ?
